common4j/src/main/com/microsoft/identity/common/java/net/UrlConnectionHttpClient.java

```diff
@@ -0,0 +1,361 @@
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.net.HttpURLConnection;
+import java.net.SocketTimeoutException;
+import java.net.URL;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.atomic.AtomicReference;
+
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.NonNull;
+
+import static com.microsoft.identity.common.java.AuthenticationConstants.AAD.CLIENT_REQUEST_ID;
+
+/**
+ * A client object for handling HTTP requests and responses.  This class accepts a RetryPolicy that
+ * is applied to the responses.  By default, the policy is a null policy not retrying anything.  The
+ * default settings for read timeout and connection timeout are 30s, and the default setting for the
+ * size of the buffer for reading objects from the http stream is 1024 bytes.
+ * <p>
+ * There are two ways to supply timeout values to this class, one method takes suppliers, the other
+ * one integers.  If you use both of these, the suppliers method will take precendence over the method
+ * using integers.
+ */
+@AllArgsConstructor
+@Builder
+@ThreadSafe
+@Immutable
+public class UrlConnectionHttpClient extends AbstractHttpClient {
+
+    protected static final int RETRY_TIME_WAITING_PERIOD_MSEC = 1000;
+
+    protected static final int STREAM_BUFFER_SIZE = 1024;
+
+    public static final int DEFAULT_CONNECT_TIME_OUT_MS = 30000;
+    public static final int DEFAULT_READ_TIME_OUT_MS = 30000;
+    public static final int DEFAULT_STREAM_BUFFER_SIZE = 1024;
+
+    @Builder.Default
+    private final IRetryPolicy<HttpResponse> retryPolicy = new NoRetryPolicy();
+    @Builder.Default
+    private final int connectTimeoutMs = DEFAULT_CONNECT_TIME_OUT_MS;
+    @Builder.Default
+    private final int readTimeoutMs = DEFAULT_READ_TIME_OUT_MS;
+    @Builder.Default
+    private final Supplier<Integer> connectTimeoutMsSupplier = null;
+    @Builder.Default
+    private final Supplier<Integer> readTimeoutMsSupplier = null;
+    @Builder.Default
+    private final int streamBufferSize = DEFAULT_STREAM_BUFFER_SIZE;
+
+    private static transient AtomicReference<UrlConnectionHttpClient> defaultReference = new AtomicReference<>(null);
+
+    public UrlConnectionHttpClient(@NonNull final UrlConnectionHttpClient client) {
+        this(client.retryPolicy,
+                client.connectTimeoutMs,
+                client.readTimeoutMs,
+                client.connectTimeoutMsSupplier,
+                client.readTimeoutMsSupplier,
+                client.streamBufferSize);
+    }
+
+    /**
+     * Obtain a static default instance of the HTTP Client class.
+     *
+     * @return a default-configured HttpClient.
+     */
+    public static synchronized UrlConnectionHttpClient getDefaultInstance() {
+        UrlConnectionHttpClient reference = defaultReference.get();
```

Comment:
Change to lazy-holder pattern?

Reply:
This is it, right? 

We're using an AtomicReference here, and we're still only init the object during the first call.